verus/lcf/README.md

```diff
@@ -0,0 +1,396 @@
+3. (in impl PartialEq for Proof) fn eq(&self, other: &Self) -> (res: bool)
+4. pub open spec fn spec_valid(self, rule_set: SpecRuleSet) -> bool
+5. pub open spec fn spec_head(self) -> SpecProp
+6. pub fn head(&self) -> (res: Prop)
+
+
+## Thm
+Theorems (thm) are used in our kernel as the final result that we are producing and want to run Ok on. If the theorem is valid, Ok will be returned, and if it is invalid, Err will be returned.
+Thm types are structs with 2 members:
+1. pub val: Prop
+2. pub p: Proof
+
+Its spec versions for SpecThm are:
+1. pub val: SpecProp
+2. pub p: SpecProof
+
+Functions that can be used on Thm types:
+1. (in impl DeepView for Thm) open spec fn deep_view(&self) -> Self::V
+2. pub open spec fn spec_wf(self, rule_set: SpecRuleSet) -> bool 
+
+## Additional Functions
+
+1. pub fn terms_eq(a: &Vec<Term>, b: &Vec<Term>) -> (res: bool)
+2. pub proof fn axiom_proof_deep_view(pf: &Proof)
+
+## mk-leaf and mk_thm
+
+mk_lf and mk_thm are called in the functions that you will write that will take take rulesets, Subst types, indexes of rules to apply, and arguments that will be used to determine if the proof tree is valid. In the toy example that will be shown, the ruleset is constructed manually by the programmer, and then the separate function testing the trace through the rules will produce a theorem if the trace is valid and return err if it is not. In the toy example and future toy examples used, it is useful to draw out the proof tree and then use mk_leaf to construct nodes of the tree that are terminal points. You can call mk_lf to call intermediate nodes and follow the trace until it's natural termination. Once the final thm has been produced, run Ok on that last thm.
+
+## Verus specific features in implementation of kernel                  
+
+One Verus specific feature in the implementation of the kernel is the separation between spec and exec code. Exec code is executable code such as functions susbt, mk_lf, and mk_thm that are commonly used in the kernel and are compileed. Spec code is code that reasons about the exec code without compiling and returns errors to the verifier if the code is not correct. Another is the use of loop invariants, triggers, assertions, and lemmas that were used to aid the verifier in reasoning about preconditions, postconditions, and recommendations. The ones left are essential for kernel verification, so do not remove them. They have been thoroughly commented and described in the code body.
+
+Another verus specific feature used was deep_view which allowed us to access the deep_view of exec functions and call spec functions in exec mode when writing invariants, preconditions, poscondtions, and asserts. Additional ensures statements were used to help the verifier understand that the deep_view call could be before or after accessing fields of dat types and also reasoning that the exec versions and spec versions functioned the same way. here's an example:
+
+    exec version:
```

Comment:
This is a long code snippet to put here---I'm not sure it is needed. For this level of detail people would just read the code.